src/mbi/relaxed_projection.py

```diff
@@ -0,0 +1,427 @@
+                     None, the optimization will start from a randomly 
+                     initialized data.
+            known_total: number of records in optimizable data (``D_prime.shape[0]``), 
+                         which can be regarded as the `batch size`.
+            optimizer: an optax optimizer, default to adam with lr=0.01.
+            seed: random seed, default to 0.
+    """
+    key = jax.random.PRNGKey(kwargs.get('seed', 0))
+
+    if isinstance(loss_fn, List):
+        # if given a list of LinearMeasurement, define the loss function by it
+        D_start = kwargs.get('D_start', None)
+        if D_start is None:
+            D_start = _initialize_synthetic_dataset(key, num_generated_points=kwargs.get('known_total', 1000), data_dimension=np.sum(domain.shape))
+
+        stat_dim = _obtain_dim(measurements = loss_fn)
+        statistics = [MarginalStatistics(domain, dim) for dim in stat_dim]
+        selected_query_index, measured_ans = _marginal_stat(loss_fn, stat_dim, statistics)
+
+        exact_fn = [None]*len(stat_dim)
+        for i in range(len(stat_dim)):
+            exact_fn[i] = statistics[i].get_exact_statistics_fn()
+
+        @jax.jit
+        def progress_loss(D_prime, query_idx, target, _exact_fn=exact_fn):
+            loss = 0.0
+            for i in range(len(stat_dim)):
+                stats = _exact_fn[i](D_prime, query_idx[i])
+                loss += jnp.linalg.norm(target[i] - stats)**2
+            return loss
+        loss_fn = jax.jit(lambda D: progress_loss(D, selected_query_index, measured_ans))
+
+        # -- initialize optimizer and update function --
+        feats_cum = jnp.array([0] + list(domain.shape)).cumsum()
+        feats_idx = [list(range(feats_cum[i], feats_cum[i+1])) for i in range(len(feats_cum)-1)]
+        optimizer = kwargs.get('optimizer', optax.adam(learning_rate=0.01))
+        opt_init_fn = optimizer.init
+
+
+        @jax.jit
+        def update_fn(D, opt_state):
+            value, grads = jax.value_and_grad(loss_fn)(D)
+            updates, opt_state = optimizer.update(grads, opt_state, D)
+            D = optax.apply_updates(D, updates)
+            D = jnp.clip(_sparsemax_project(D, feats_idx), 0, 1)
+            return D, opt_state, value
+
+
+        # -- main optimization step: find a D' that minimized the loss function -- 
+        progress_loss_start = loss_fn(D_start)
+        D_prime = _optimize_D(
+            D_start,
+            opt_init_fn,
+            loss_fn,
+            update_fn,
+            iters
+        )
+        progress_loss_final = loss_fn(D_prime)
+        print(
+            f'marginal fitting results: start loss = {progress_loss_start}, end loss = {progress_loss_final}'
+        )
+
+        return ProjectableData(D_prime = D_prime, domain = domain)
+    
+    else:
+        # if given a MarginalLossFn, we transform D_start into CliqueVector and optimize it (same as PGM)
```

Comment:
This "else block" looks good to me.  Let's go through this codepath even when given a list of LinearMeasurements, by using this function https://github.com/ryan112358/mbi/blob/master/src/mbi/marginal_loss.py#L100